File: expo-sdk/ios/MdocDataTransferModule.swift
import ExpoModulesCore
import MdocDataModel18013
import SwiftCBOR

public class MdocDataTransferModule: Module {
  public func definition() -> ModuleDefinition {
    Name("MdocDataTransfer")
      
    let bleServerTransfer: MdocGattServer = MdocGattServer(self)
      
    Events(
        MdocDataTransferEvent.ON_REQUEST_RECEIVED.description,
        MdocDataTransferEvent.ON_RESPONSE_SENT.description
    )

    Function("initialize") {
        // no-op
    }
      
      AsyncFunction("startQrEngagement") {
          try await bleServerTransfer.performDeviceEngagement()
      }
      
      AsyncFunction("sendDeviceResponse") { (deviceResponse: String) in
        try await bleServerTransfer.sendDeviceResponse(deviceResponse: Data(base64Encoded: deviceResponse.data(using: .utf8)!)!)
        sendEvent(MdocDataTransferEvent.ON_RESPONSE_SENT.description)
    }
      
    Function("shutdown") {
        bleServerTransfer.stop()
    }
      
    Function("enableNfc") {
        // no-op
    }
  }
}

extension MdocDataTransferModule : MdocOfflineDelegate {
    public func didChangeStatus(_ newStatus: TransferStatus) {
        // no-op
    }
    
    public func didFinishedWithError(_ error: any Error) {
        // no-op
    }
    
    public func didReceiveRequest(_ deviceRequest: Data, _ sessionTranscript: Data) {
        do {
            let dr = try DeviceRequest(data: deviceRequest.bytes).encode(options: CBOROptions())
            sendEvent(MdocDataTransferEvent.ON_REQUEST_RECEIVED.description, ["deviceRequest": Data(dr).base64EncodedString(),"sessionTranscript": sessionTranscript.base64EncodedString() ])
        } catch {
            logger.error("\(error.localizedDescription)")
        }
    }
    
    
}

File: expo-sdk/android/src/main/java/id/animo/mdoc/datatransfer/MdocDataTransferModule.kt
package foundation.openwallet.mdoc.datatransfer

import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi

class MdocDataTransferModule : Module() {
    @OptIn(ExperimentalEncodingApi::class)
    override fun definition() = ModuleDefinition {
        var mDocDataTransfer: MdocDataTransfer? = null

        Name("MdocDataTransfer")

        Events(
            MdocDataTransferEvent.ON_REQUEST_RECEIVED,
            MdocDataTransferEvent.ON_RESPONSE_SENT,
            MdocDataTransferEvent.ON_ERROR
        )

        Function("initialize") {
            mDocDataTransfer = MdocDataTransfer(
                appContext.reactContext ?: throw Exceptions.ReactContextLost(),
                appContext.currentActivity ?: throw Exceptions.MissingActivity()
            ) { name: String, body: Map<String, Any?>? ->
                sendEvent(
                    name,
                    body ?: mapOf()
                )
            }

            return@Function null
        }


        AsyncFunction("startQrEngagement") { promise: Promise ->
            val transfer = mDocDataTransfer ?: throw MdocDataTransferException.NotInitialized()

            // Whichever callback fires first settles the promise and clears both, so a later event
            // cannot settle it a second time.
            fun clearCallbacks() {
                transfer.onQrEngagementReady = null
                transfer.onEngagementError = null
            }

            transfer.onQrEngagementReady = { qrCode ->
                clearCallbacks()
                promise.resolve(qrCode)
            }
            transfer.onEngagementError = { error ->
                clearCallbacks()
                promise.reject(MdocDataTransferException.EngagementFailed(error))
            }

            transfer.startQrEngagement()
        }

        Function("sendDeviceResponse") { deviceResponse: String ->
            (mDocDataTransfer ?: throw MdocDataTransferException.NotInitialized()).respond(
                Base64.Default.decode(deviceResponse)
            )
        }

        Function("shutdown") {
            (mDocDataTransfer ?: throw MdocDataTransferException.NotInitialized()).shutdown()
        }

        Function("enableNfc") {
            (mDocDataTransfer ?: throw MdocDataTransferException.NotInitialized()).enableNfc()
        }
    }
}
